wgengine/userspace.go

```diff
@@ -1295,6 +1311,11 @@ func (e *userspaceEngine) SetDERPMap(dm *tailcfg.DERPMap) {
 
 func (e *userspaceEngine) SetNetworkMap(nm *controlclient.NetworkMap) {
 	e.magicConn.SetNetworkMap(nm)
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	for _, fn := range e.networkMapCallbacks {
+		fn(nm)
```

Comment:
so now the potential worry is that these callbacks are held with e.mu held, which means if they try to do a call to the Engine directly or indirectly, we deadlock.

that's one of the dozen issues with callback-based APIs you have to worry about & document.

one thing we could do here is copy the callback(s) out of the engine, release the lock, and then call all the callbacks.